### PR TITLE
feat: add timeout for gen proof

### PR DESCRIPTION
**Pull request by johntaiko**

### CODE REVIEW

File: prover/src/shared_state.rs [hunk 1]
```diff
@@ -559,11 +569,14 @@ impl SharedState {
...
-            let pk = Arc::new(pk);
-
+            let param = param.clone();
+            let circuit = circuit.clone();
+            let handle = tokio::task::spawn_blocking(move || {
+                let vk = keygen_vk(param.as_ref(), &circuit)?;
+                let pk = keygen_pk(param.as_ref(), vk, &circuit)?;
+                Result::<_, halo2_proofs::plonk::Error>::Ok(Arc::new(pk))
+            });
+            let pk = tokio::time::timeout(GEN_TIMEOUT, handle).await???;
```
Comment (smtmfft) [inline prover/src/shared_state.rs:588]: why 3 ? here?

File: prover/src/shared_state.rs [hunk 2]
```diff
@@ -189,21 +191,29 @@ macro_rules! gen_proof {
...
+                        circuit_instance.clone(),
+                        fixed_rng(),
+                        task_options.mock_feedback,
+                        task_options.verify_proof,
+                    )
+                });
+
+                let proof = tokio::time::timeout(GEN_TIMEOUT, handle)
+                    .await
+                    .map_err(|e| e.to_string())?
```
Comment (smtmfft) [inline prover/src/shared_state.rs]: what does the error string look like? If david want to query in log.

Comment (johntaiko) [reply prover/src/shared_state.rs]: one of four:
1. "deadline has elapsed" when timeout
2. "task {id} was cancelled" when spawned gen task be cancelled
3. "task {id} panicked" when spawned gen task panic
4. other error from gen task

File: prover/src/shared_state.rs (hunk 1 above)
Comment (johntaiko) [reply prover/src/shared_state.rs:588]: TimeoutError(spawnError(taskError)))

**Comment by smtmfft** [comment]
Please change to 15mins, then it will be merged.

**Pull request merged**